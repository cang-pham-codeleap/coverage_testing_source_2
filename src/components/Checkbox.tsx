import React, { useState } from "react";

interface CheckboxProps {
  /**
   * Whether the checkbox is checked
   */
  checked?: boolean;

  /**
   * Default checked state (for uncontrolled component)
   */
  defaultChecked?: boolean;

  /**
   * Disabled state of the checkbox
   */
  disabled?: boolean;

  /**
   * Function called when the checkbox state changes
   */
  onChange?: (checked: boolean) => void;

  /**
   * Size of the checkbox: 'sm' (small), 'md' (medium), or 'lg' (large)
   */
  size?: "sm" | "md" | "lg";

  /**
   * Label text to display next to the checkbox
   */
  label?: string;

  /**
   * Additional CSS class names
   */
  className?: string;

  /**
   * ID for the checkbox input
   */
  id?: string;

  /**
   * Name attribute for the checkbox input
   */
  name?: string;

  /**
   * Value attribute for the checkbox input
   */
  value?: string;

  /**
   * Required attribute
   */
  required?: boolean;

  /**
   * Indeterminate state (partially checked)
   */
  indeterminate?: boolean;
}

/**
 * Checkbox component for selecting options
 *
 * @example
 * ```tsx
 * <Checkbox label="Accept terms" onChange={handleChange} />
 * <Checkbox checked={isSelected} onChange={setIsSelected} label="Select option" />
 * <Checkbox indeterminate size="lg" label="Some items selected" />
 * ```
 */
const Checkbox: React.FC<CheckboxProps> = ({
  checked,
  defaultChecked = false,
  disabled = false,
  onChange,
  size = "md",
  label,
  className = "",
  id,
  name,
  value,
  required = false,
  indeterminate = false,
}) => {
  // For uncontrolled component usage
  const [internalChecked, setInternalChecked] =
    useState<boolean>(defaultChecked);

  // Determine if component is controlled or uncontrolled
  const isControlled = checked !== undefined;
  const isChecked = isControlled ? checked : internalChecked;

  const checkboxId = React.useId();

  // Generate unique ID if not provided
  const uniqueId = React.useMemo(() => id || checkboxId, [id]);

  // Size-based styling
  const sizeClasses = {
    sm: {
      checkbox: "w-3 h-3",
      label: "text-sm",
      checkmark: "w-2 h-2",
    },
    md: {
      checkbox: "w-4 h-4",
      label: "text-base",
      checkmark: "w-3 h-3",
    },
    lg: {
      checkbox: "w-5 h-5",
      label: "text-lg",
      checkmark: "w-4 h-4",
    },
  };

  // Handle change
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    // Update internal state for uncontrolled usage
    if (!isControlled) {
      setInternalChecked(e.target.checked);
    }

    // Call external onChange handler if provided
    if (onChange) {
      onChange(e.target.checked);
    }
  };

  // Set indeterminate property using ref
  const checkboxRef = React.useRef<HTMLInputElement>(null);

  React.useEffect(() => {
    if (checkboxRef.current) {
      checkboxRef.current.indeterminate = indeterminate;
    }
  }, [indeterminate]);

  return (
    <div className={`flex items-center ${className}`}>
      <div className="relative flex items-center">
        <input
          ref={checkboxRef}
          type="checkbox"
          id={uniqueId}
          name={name}
          value={value}
          checked={isChecked}
          onChange={handleChange}
          disabled={disabled}
          required={required}
          className={`
            ${sizeClasses[size].checkbox}
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            ${isChecked ? "bg-blue-600 border-blue-600" : ""}
            ${indeterminate ? "bg-blue-600 border-blue-600" : ""}
            ${disabled ? "opacity-50 cursor-not-allowed" : ""}
          `}
          aria-checked={indeterminate ? "mixed" : isChecked}
        />
        {/* Custom checkmark */}
        {isChecked && !indeterminate && (
          <div className="absolute pointer-events-none text-white flex items-center justify-center">
            <svg
              className={sizeClasses[size].checkmark}
              viewBox="0 0 16 16"
              fill="none"
              stroke="currentColor"
              strokeWidth="3"
              strokeLinecap="round"
              strokeLinejoin="round"
            >
              <path d="M4 8.5L6.5 11L12 5" />
            </svg>
          </div>
        )}
        {/* Indeterminate state dash */}
        {indeterminate && (
          <div className="absolute pointer-events-none text-white flex items-center justify-center">
            <svg
              className={sizeClasses[size].checkmark}
              viewBox="0 0 16 16"
              fill="none"
              stroke="currentColor"
              strokeWidth="3"
              strokeLinecap="round"
              strokeLinejoin="round"
            >
              <path d="M4 8L12 8" />
            </svg>
          </div>
        )}
      </div>
      {label && (
        <label
          htmlFor={uniqueId}
          className={`ml-2 ${sizeClasses[size].label} ${
            disabled ? "opacity-50 cursor-not-allowed" : "cursor-pointer"
          }`}
        >
          {label}
        </label>
      )}
    </div>
  );
};

export default Checkbox;
