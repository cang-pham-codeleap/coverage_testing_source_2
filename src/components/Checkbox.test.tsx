import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Checkbox from "./Checkbox";

describe("Checkbox Component", () => {
  // Test rendering with default props
  it("renders correctly with default props", () => {
    const { container } = render(<Checkbox />);
    expect(container).toMatchSnapshot();
  });

  // Test with label
  it("renders with a label", () => {
    render(<Checkbox label="Test Label" />);
    expect(screen.getByText("Test Label")).toMatchSnapshot();
    expect(screen.getByLabelText("Test Label")).toMatchSnapshot();
  });

  // Test controlled state changes
  it("handles controlled state changes", () => {
    const handleChange = jest.fn();
    const { rerender } = render(
      <Checkbox checked={false} onChange={handleChange} />
    );

    // Find the input and click it
    const input = screen.getByRole("checkbox");
    fireEvent.click(input);

    // Check if onChange was called with true
    expect(handleChange).toHaveBeenCalledWith(true);

    // Rerender with checked=true to simulate controlled component update
    rerender(<Checkbox checked={true} onChange={handleChange} />);
    expect(input).toMatchSnapshot();
  });

  // Test uncontrolled state behavior
  it("handles uncontrolled state changes", () => {
    render(<Checkbox defaultChecked={false} />);

    const input = screen.getByRole("checkbox");
    expect(input).toMatchSnapshot();

    // Click the checkbox
    fireEvent.click(input);
    expect(input).toMatchSnapshot();

    // Click again to toggle off
    fireEvent.click(input);
    expect(input).toMatchSnapshot();
  });

  // Test disabled state
  it("disables the checkbox when disabled prop is true", () => {
    const handleChange = jest.fn();
    render(<Checkbox disabled onChange={handleChange} />);

    const input = screen.getByRole("checkbox");
    expect(input).toMatchSnapshot();

    // Try to click the disabled checkbox
    fireEvent.click(input);
    expect(handleChange).toMatchSnapshot();
  });

  // Test indeterminate state
  it("sets the indeterminate state correctly", () => {
    render(<Checkbox indeterminate />);

    const input = screen.getByRole("checkbox");
    expect(input).toMatchSnapshot();

    // We can't directly check the indeterminate property through testing-library
    // because it's a DOM property, not an attribute
  });

  // Test different sizes
  it("renders with different sizes", () => {
    const { rerender, container } = render(
      <Checkbox size="sm" label="Small" />
    );
    expect(container).toMatchSnapshot();

    rerender(<Checkbox size="md" label="Medium" />);
    expect(container).toMatchSnapshot();

    rerender(<Checkbox size="lg" label="Large" />);
    expect(container).toMatchSnapshot();
  });

  // Test with custom props
  it("passes through custom props", () => {
    render(
      <Checkbox
        id="custom-id"
        name="custom-name"
        value="custom-value"
        required
      />
    );

    const input = screen.getByRole("checkbox");
    expect(input).toMatchSnapshot();
  });

  // Test with custom class name
  it("applies custom class name", () => {
    render(<Checkbox className="custom-class" />);
    const wrapper = screen.getByRole("checkbox").closest("div");
    expect(wrapper).toMatchSnapshot();
  });
});
