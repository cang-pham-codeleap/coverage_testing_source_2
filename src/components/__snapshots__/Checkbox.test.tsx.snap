// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component applies custom class name 1`] = `
<div
  class="relative flex items-center"
>
  <input
    aria-checked="false"
    class="
            w-4 h-4
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            
            
            
          "
    id="checkbox-0ljx9umlt"
    type="checkbox"
    value=""
  />
</div>
`;

exports[`Checkbox Component disables the checkbox when disabled prop is true 1`] = `
<input
  aria-checked="false"
  class="
            w-4 h-4
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            
            
            opacity-50 cursor-not-allowed
          "
  disabled=""
  id="checkbox-8mnm81qro"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox Component disables the checkbox when disabled prop is true 2`] = `
[MockFunction] {
  "calls": [
    [
      true,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Checkbox Component handles controlled state changes 1`] = `
<input
  aria-checked="true"
  class="
            w-4 h-4
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            bg-blue-600 border-blue-600
            
            
          "
  id="checkbox-b5vzccnzt"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox Component handles uncontrolled state changes 1`] = `
<input
  aria-checked="false"
  class="
            w-4 h-4
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            
            
            
          "
  id="checkbox-u7qxsj3cq"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox Component handles uncontrolled state changes 2`] = `
<input
  aria-checked="true"
  class="
            w-4 h-4
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            bg-blue-600 border-blue-600
            
            
          "
  id="checkbox-u7qxsj3cq"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox Component handles uncontrolled state changes 3`] = `
<input
  aria-checked="false"
  class="
            w-4 h-4
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            
            
            
          "
  id="checkbox-u7qxsj3cq"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox Component passes through custom props 1`] = `
<input
  aria-checked="false"
  class="
            w-4 h-4
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            
            
            
          "
  id="custom-id"
  name="custom-name"
  required=""
  type="checkbox"
  value="custom-value"
/>
`;

exports[`Checkbox Component renders correctly with default props 1`] = `
<div>
  <div
    class="flex items-center "
  >
    <div
      class="relative flex items-center"
    >
      <input
        aria-checked="false"
        class="
            w-4 h-4
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            
            
            
          "
        id="checkbox-cyxd3b4ui"
        type="checkbox"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`Checkbox Component renders with a label 1`] = `
<label
  class="ml-2 text-base cursor-pointer"
  for="checkbox-ee1jmcjag"
>
  Test Label
</label>
`;

exports[`Checkbox Component renders with a label 2`] = `
<input
  aria-checked="false"
  class="
            w-4 h-4
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            
            
            
          "
  id="checkbox-ee1jmcjag"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox Component renders with different sizes 1`] = `
<div>
  <div
    class="flex items-center "
  >
    <div
      class="relative flex items-center"
    >
      <input
        aria-checked="false"
        class="
            w-3 h-3
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            
            
            
          "
        id="checkbox-9eehmzvb6"
        type="checkbox"
        value=""
      />
    </div>
    <label
      class="ml-2 text-sm cursor-pointer"
      for="checkbox-9eehmzvb6"
    >
      Small
    </label>
  </div>
</div>
`;

exports[`Checkbox Component renders with different sizes 2`] = `
<div>
  <div
    class="flex items-center "
  >
    <div
      class="relative flex items-center"
    >
      <input
        aria-checked="false"
        class="
            w-4 h-4
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            
            
            
          "
        id="checkbox-9eehmzvb6"
        type="checkbox"
        value=""
      />
    </div>
    <label
      class="ml-2 text-base cursor-pointer"
      for="checkbox-9eehmzvb6"
    >
      Medium
    </label>
  </div>
</div>
`;

exports[`Checkbox Component renders with different sizes 3`] = `
<div>
  <div
    class="flex items-center "
  >
    <div
      class="relative flex items-center"
    >
      <input
        aria-checked="false"
        class="
            w-5 h-5
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            
            
            
          "
        id="checkbox-9eehmzvb6"
        type="checkbox"
        value=""
      />
    </div>
    <label
      class="ml-2 text-lg cursor-pointer"
      for="checkbox-9eehmzvb6"
    >
      Large
    </label>
  </div>
</div>
`;

exports[`Checkbox Component sets the indeterminate state correctly 1`] = `
<input
  aria-checked="mixed"
  class="
            w-4 h-4
            appearance-none
            border border-gray-300
            rounded
            bg-white
            transition-colors
            duration-200
            cursor-pointer
            focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
            
            bg-blue-600 border-blue-600
            
          "
  id="checkbox-kii06k660"
  type="checkbox"
  value=""
/>
`;
